hypothesis-python/src/hypothesis/core.py

```diff
@@ -996,7 +996,7 @@ def _execute_once_for_engine(self, data: ConjectureData) -> None:
         except UnsatisfiedAssumption:
             # An "assume" check failed, so instead we inform the engine that
             # this test run was invalid.
-            data.mark_invalid()
+            data.mark_invalid("failed to satisfy assume()")
```

Comment:
This explicitly mentions `assume()`, which isn't technically correct. `UnsatisfiedAssumption` might be raised by e.g. `reject()`. 

But given that `reject()` seems to be soft-deprecated in favor of `assume(False)`, and `assume()` is the vast majority of usages anyway, this felt a bit better than the more vague "failed to satisfy some assumption", which could be used to describe almost any usage of `mark_invalid`! I'm on the fence though.

Reply:
I definitely don't want to leave potentially-misleading notes!  Here's an idea:

- Give `UnsatisfiedAssumption` a `reason` kwarg/attribute
- In `assume()` and `reject()`, pass the reason string.  
  - For bonus points, use [`inspect.stack(context=2)`](https://docs.python.org/3/library/inspect.html#inspect.stack) to work out where it was called from.
- If we don't have this information, don't add the event.

Reply:
Something like this 9dbfbc7111062d6cdbef2d3d88f4c5e3c25233e3? I assume you meant `inspect.stack()[1]`, because `context` increases the amount of source code lines returned around that stackframe. We *could* use that to work out the exact call site of the `assume`, instead of just its calling function, but then multiline `assume` statements become a problem...